File: ios/WatchApp Watch App/Models/GameNode.swift
import Foundation

/// A class representing a single node in the "Contact Abyss" game.
///
/// Each `GameNode` contains the narrative text, sensor data, available choices, battery power changes,
/// and information about whether the node signifies the end of the game.
class GameNode {
    /// A unique identifier for the node.
    let id: String
    
    /// The narrative text displayed to the player at this node.
    let storyText: String
    
    /// Optional sensor data associated with this node.
    let sensorData: SensorData?
    
    /// A list of choices available to the player at this node.
    let choices: [Choice]
    
    /// The change in battery power resulting from the player's decision at this node.
    let batteryChange: Int?
    
    /// Indicates whether this node is an ending point of the game.
    let isEnd: Bool
    
    /// Determines if the game outcome is a win, loss, or neutral.
    /// A `GameOutcome` will only be available for ending nodes.
    let outcome: GameOutcome?
    
    /// Creates a new instance of `GameNode`.
    ///
    /// - Parameters:
    ///   - id: A unique identifier for the node.
    ///   - storyText: The narrative text displayed to the player.
    ///   - sensorData: Optional sensor data related to the node.
    ///   - choices: A list of available choices at this node.
    ///   - batteryChange: Optional change in battery power.
    ///   - isEnd: Flag indicating if this node is an ending point.
    ///   - outcome: Optional game outcome associated with this node.
    init(
        id: String,
        storyText: String,
        sensorData: SensorData?,
        choices: [Choice],
        batteryChange: Int?,
        isEnd: Bool = false,
        outcome: GameOutcome? = nil
    ) {
        self.id = id
        self.storyText = storyText
        self.sensorData = sensorData
        self.choices = choices
        self.batteryChange = batteryChange
        self.isEnd = isEnd
        self.outcome = outcome
    }
    
    /// Creates a `GameNode` instance from a JSON dictionary.
    ///
    /// - Parameter json: A `[String: Any]` dictionary containing the node's data.
    /// - Returns: A new instance of `GameNode` or `nil` if the dictionary is invalid.
    convenience init?(json: [String: Any]) {
        guard let id = json["id"] as? String,
              let storyText = json["story_text"] as? String else {
            return nil
        }
        
        let sensorData = (json["sensor_data"] as? [String: Any]).flatMap { SensorData(json: $0) }
        
        let choices = (json["choices"] as? [[String: Any]] ?? []).compactMap { Choice(json: $0) }
        
        let batteryChange = json["battery_change"] as? Int
        
        let isEnd = json["is_end"] as? Bool ?? false
        
        let outcomeString = json["outcome"] as? String
        let outcome = outcomeString.flatMap { GameOutcome(rawValue: $0) }
        
        self.init(
            id: id,
            storyText: storyText,
            sensorData: sensorData,
            choices: choices,
            batteryChange: batteryChange,
            isEnd: isEnd,
            outcome: outcome
        )
    }
    
    /// Converts the `GameNode` instance into a JSON dictionary.
    ///
    /// - Returns: A `[String: Any]` dictionary representing the node's data.
    func toJson() -> [String: Any] {
        var json: [String: Any] = [
            "id": id,
            "story_text": storyText,
            "choices": choices.map { $0.toJson() },
            "is_end": isEnd
        ]
        
        if let sensorData = sensorData {
            json["sensor_data"] = sensorData.toJson()
        }
        
        if let batteryChange = batteryChange {
            json["battery_change"] = batteryChange
        }
        
        if let outcome = outcome {
            json["outcome"] = outcome.rawValue
        }
        
        return json
    }
}

File: lib/services/game_service/models/game_node.dart
import 'package:contact_abyss/services/game_service/models/choice.dart';
import 'package:contact_abyss/services/game_service/models/game_outcome.dart';
import 'package:contact_abyss/services/game_service/models/sensor_data.dart';

/// A class representing a single node in the "Contact Abyss" game.
///
/// Each `GameNode` contains the narrative text, sensor data, available choices, battery power changes, and
/// information about whether the node signifies the end of the game. This structure allows for the creation of an
/// n-ary tree that models the game's decision-making process.
class GameNode {
  /// A unique identifier for the node.
  final String id;

  /// The narrative text displayed to the player at this node.
  final String storyText;

  /// Optional sensor data associated with this node.
  ///
  /// This may include information such as battery level, radiation levels,
  /// and other environmental readings that influence the player's decisions.
  final SensorData? sensorData;

  /// A list of choices available to the player at this node.
  ///
  /// Each choice leads to another `GameNode` identified by its target ID.
  final List<Choice> choices;

  /// The change in battery power resulting from the player's decision at this node.
  ///
  /// A positive value indicates a gain in battery power, while a negative value
  /// signifies a loss. This helps manage the battery constraints within the game.
  final int? batteryChange;

  /// Indicates whether this node is an ending point of the game.
  ///
  /// If `true`, the game concludes upon reaching this node.
  final bool isEnd;

  /// Determines if the game outcome is a win, loss, or neutral. A [GameOutcome] will only be available for ending
  /// nodes.
  final GameOutcome? outcome;

  /// Creates a new instance of [GameNode].
  ///
  /// - [id]: A unique identifier for the node.
  /// - [storyText]: The narrative text displayed to the player.
  /// - [choices]: A list of available choices at this node.
  /// - [sensorData]: Optional sensor data related to the node.
  /// - [batteryChange]: Optional change in battery power.
  /// - [isEnd]: Optional flag indicating if this node is an ending point.
  GameNode({
    required this.id,
    required this.storyText,
    required this.choices,
    this.sensorData,
    this.batteryChange,
    this.isEnd = false,
    this.outcome,
  });

  /// Creates a [GameNode] instance from a JSON map.
  ///
  /// This factory constructor is useful for deserializing JSON data into
  /// `GameNode` objects.
  ///
  /// - [json]: A `Map<String, dynamic>` containing the node's data.
  ///
  /// Returns a new instance of [GameNode].
  factory GameNode.fromJson(Map<String, dynamic> json) {
    return GameNode(
      id: json['id'] as String,
      storyText: json['story_text'] as String,
      sensorData: json['sensor_data'] != null ? SensorData.fromJson(json['sensor_data'] as Map<String, dynamic>) : null,
      choices: (json['choices'] as List<dynamic>? ?? [])
          .map((choice) => Choice.fromJson(choice as Map<String, dynamic>))
          .toList(),
      batteryChange: json['battery_change'] != null ? json['battery_change'] as int : null,
      isEnd: json['is_end'] as bool? ?? false,
      outcome: json['outcome'] != null
          ? GameOutcome.values.firstWhere((outcome) => outcome.id == json['outcome'] as String)
          : null,
    );
  }

  /// Converts the [GameNode] instance into a JSON map.
  ///
  /// This method is useful for serializing `GameNode` objects into JSON format.
  ///
  /// Returns a `Map<String, dynamic>` representing the node's data.
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'story_text': storyText,
      if (sensorData != null) 'sensor_data': sensorData!.toJson(),
      'choices': choices.map((choice) => choice.toJson()).toList(),
      if (batteryChange != null) 'battery_change': batteryChange,
      'is_end': isEnd,
    };
  }
}
